[MDB IGNORE] TGUI library, plus polish

LemonInTheDark opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

This is a recreation of #62509
I've fixed up some stuff I've been meaning to for a bit 
I've also done some polishing, giving the different machines audio feedback and stuff.

<details> <summary>Overarching behavior stuff </summary>

Updates all three library machines to tgui. So that's the visitor console, the book management console, and the scanner.
Adds some helper datums to make passing book information around easier
Adds hard limits on book title/author/contents
Reworks how bookshelves load books in. Instead of loading lazily, we load them during init in a separate subsystem. This opens us up to being able to do fun things, like autofilling a library console's inventory page. 
Adds a map config option for what areas count as part of "a library" in the eyes of the above system so kilo isn't woefully underfilled.
You can now choose what type of poster to print in the console, rather then getting a random item.
You can now choose to heed the warnings of the eldritch software company.
You can see what type of bible the chaplin has selected/who he worships?
You can no longer infinitely dump books into the library scanner.
</details>

<details> <summary>Implementation details, BYOND </summary>
Any input that makes a db request will now A: freeze up any other db inputs until it's finished, and B: Start a 1 second timer before any new db requests can be made. You're also not allowed to make the same search request twice.
Of note, I'm handling html encoding in a very targeted way.
All book_data datums need to have html encoded values. get_title/author/content exist so a defaulting and tgui appropriate version can be loaded in. This somewhat matches with the trusted var on set_title, it exists to prevent double html encoding.

Looks like this
Input/DB (Book data should be html encoded)
Inside book datum (Book data should be html encoded)
Sending to tgui (Book data should be decoded during extraction with the get_() procs)
Sending anywhere else (Book data should be html encoded, otherwise it's an xss vuln)
</details>

<details> <summary>Implementation details, TGUI</summary>

The visitor and book management console share the same data/act pipeline, which is why they're parented/subtyped
They also share a page selection component, which is why the visitor's console imports it.
I'm using a custom theme for emag visuals
The page selection component is kinda cursed, the left and right controls are fine, but I'm trying to get a `<< < [page/max] > >>` setup going, and that means resetting the center input after every change so the placeholder can be used
This ends up being slightly hacky. I'm sorry.

I've added a custom theme for the library console when it's emagged. There's a few semi cursed things about this.
1: I'm overriding a name to color mapping list in an odd way, since I'm pretty sure it's precomputed. Might not be needed though, I'm not the most comfortable with scss.
2: I added a new svg background for the ui, because I thought it would be a nice touch. I'm very out of my depth with this. It's currently presenting two problems.
One, is it even worth having in the first place. I don't have a very good concept of cost for uis like this, so if any part of this is needless bloat I'll just nuke it.
Two, how the hell do you compress these things. I've tried online options, and they'll display in my editor, but when I try and use them in the theme itself they just crap out. It's called -raw for a reason, I realize the current size is nowhere near acceptable. On that note, may redo the thing with less spirals.
As a part of all that I've added a dimness property to dimmer scss so themes can control how dark things get. Need this cause otherwise things just look washed out when modals are displayed.
 
Oh also, I implemented a custom tab setup for this ui. I have no idea why it was literally like 5 months ago.
I think it looks pretty nice, but if you want me to nuke it I can. Sorry for any headache around this.

One more thing.
For one of the pages in the ui, I'm displaying html raw. This is so markdownified books look goodish. 
I'm sanitizing them of course, but I ran into what I think are holes in our current whitelist. I think these are fine, can't think of any problems extra bold, strikethrough, underline and accent could cause, but I'm me so I wait to be proved an idiot.

Sorry if I missed anything here, this has been a few weeks of nightcode and I'm kinda running on fumes at the moment
</details>

### Pictures 
<details> <summary>Library Scanner </summary>

![image](https://user-images.githubusercontent.com/58055496/156997394-60249db0-4888-47ad-b1d7-751355dfb8ac.png)
</details>
<details> <summary>
Visitor's Console
</summary>

![image](https://user-images.githubusercontent.com/58055496/139661344-8da04435-6eca-4dcf-a3ee-47b967018063.png)
</details>
<details> <summary>
Inventory Page
</summary>

![image](https://user-images.githubusercontent.com/58055496/139661373-3587c042-f3b8-46e0-b8cf-f58e81d7736d.png)
</details>
<details> <summary>
Checkout Manager
</summary>

![image](https://user-images.githubusercontent.com/58055496/139661412-c60af584-8d71-4602-85ef-098e10b5a654.png)
</details>
<details> <summary>
Checkout Popup
</summary>

![image](https://user-images.githubusercontent.com/58055496/139661434-84744f77-f352-42c3-a434-94e47fa77fd2.png)
</details>
<details> <summary>
Archive/DB search
</summary>

![image](https://user-images.githubusercontent.com/58055496/139661473-a48b53d1-9593-49f3-83c5-5192db82f5eb.png)
</details>
<details> <summary>
Upload to Newscaster/DB + Scanner Cache preview
</summary>

![image](https://user-images.githubusercontent.com/58055496/139661517-3aaa479a-f14b-4773-a098-9492aa86137e.png)
</details>
<details> <summary>
Upload to DB popup
</summary>

![image](https://user-images.githubusercontent.com/58055496/139812620-8236debb-5a09-4ce1-b00f-56315bc9f4e0.png)
</details>
<details> <summary>
Misc Printing, Bibles and Posters
</summary>

![image](https://user-images.githubusercontent.com/58055496/139661594-fc575edf-7c4e-4d90-b119-8e717774249a.png)
</details>
<details> <summary>
Collapsed Tabs
</summary>

![image](https://user-images.githubusercontent.com/58055496/139661659-d1c8af5f-6fe9-4c7b-b58c-8bf539884105.png)
</details>
<details> <summary>
Emag Theme + Forbidden Page
</summary>

![image](https://user-images.githubusercontent.com/58055496/139661684-54a017db-02b2-4a05-81dc-b0bd52988a84.png)
</details>
<details> <summary>
Scanner Cache preview with html tags
</summary>

![image](https://user-images.githubusercontent.com/58055496/139661764-0c9a140a-67f2-4fb3-8e7c-d037dd9ced3f.png)
</details>

## Why It's Good For The Game

Haha user interface go brrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrr
TODO:

- [x] Audit input box length, see if you can increase them to better match the size limits on titles and author pages
- [x] Check to see what markdown required tags exist, add them
- [x] Lower the size of that background
- [x] Commit and push the remaining map repaths, lost in the sauce of my shitty pc
- [ ] Don't kill stylemistake
- [x] Default Bible Name
- [x] Test print page with max length everything, make sure it's ok
- [x] Add emissives to consoles

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. -->

:cl:
expansion: The library computers have received a graphical and functional upgrade. Better inventory management, poster printing, bible viewing, and well, TGUI.
server: Bookcases will generate way more db calls during init then is normal. If this causes issues please bug me and I'll work out a solution
fix: Library consoles and a few other things will no longer look "on" even when their power is cut.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

LemonInTheDark committed:
Prevents potentially infinite length books from being written and stored. I'm not sure if this is an actual issue, but I have a funny feeling it may become one someday

LemonInTheDark committed:
Moves the paper defines to their own file

LemonInTheDark committed:
It's become clear to me that I am stupid

LemonInTheDark committed:
git add --all

LemonInTheDark committed:
Makes book info into a datum to allow for easy passing around

LemonInTheDark committed:
Converts the library scanner to tgui, lays the groundwork for tgui visitor consoles

LemonInTheDark committed:
Makes the db request for book info sort
Adds the frontend for the visitor's console
Adds a hash to prevent duplicate db requests
Adds a prams changed var to help facilitate a better search button
Makes the page number code accept text as input

LemonInTheDark committed:
Makes the ui index at 1 even tho we index at 0 internally

LemonInTheDark committed:
Begins the conversion process for the library console.

Changes the library console to override the visitor, to utalize for the archive access portion of the ui
Makes scanner into a weakref, I'm coming for you handheld scanner
Renames some vars to make things clearer

LemonInTheDark committed:
Converts the remaining refs of the old console typepath over, adds a circuit board for consoles because pain

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into book

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into book

LemonInTheDark committed:
Changes how bookshelves load in books
Instead of loading them in lazyally, we load them during init
This lets us track what books are stored in which areas

Somewhat jutting off of this, adds map config for designating something as "part of the library"
This will be useful later

LemonInTheDark committed:
Renames the random poster, adds a spritesheet for bibles. Both will be useful in a moment

LemonInTheDark committed:
Ok. This is a bit of a mess.
Converts the library console to tgui.
This comes with a few minor behavior changes:
You can now select what type of poster you want to print, instead of just printing a random one
It's now possible to heed the console's emag warning
The console's inventory page will fill at roundstart with the books in your area/if you're in a library, any
areas designated as "library like" in the map config
You can see what type of bible the chaplin has selected?

"Fixes":
You can no longer just dump books into the scanner forever

Implementation details:
Any input that makes a db request will now A: freeze up any other db inputs until it's finished, and B: Start a
1 second timer before any new db requests can be made
Of note, I'm handling html encoding in a very targeted way.
All book_data datums need to have html encoded values. get_title/author/content exist so a defaulting and tgui
appropriate version can be loaded in. This somewhat matches with the trusted var on set_title, it exists to
prevent double html encoding.

While we're here

Input/DB (Book data should be html encoded)
Inside book datum (Book data should be html encoded)
Sending to tgui (Book data should be decoded during extraction with the get_() procs)
Sending anywhere else (Book data should be html encoded, otherwise it's an xss vuln)

Uhhhh tgui stuff?
I'm using a custom theme for emag visuals, I'll get into that more later
The visitor and book management console share the same data/act pipeline, which is why they're parented/subtyped
They also share a page selection component, which is why the visitor's console imports it.

Uhhhhh
Oh right, fuck.
Ok so the page selection component is kinda cursed, the left and right controls are fine
But I'm trying to get a << < [page/max] > >> setup going, and that means resetting the center input past change
so the default value can be used
This ends up being slightly hacky. I'm sorry.

Oh also, I implemented a custom tab setup for this ui. I have no idea why it was literally like 5 months ago.
I think it looks pretty nice, but if you want me to nuke it I can. Sorry for any headache around this.

More tgui stuff next

LemonInTheDark committed:
Scanner/visitor cleanup, some other odds and ends

LemonInTheDark committed:
Adds in a dark red and black theme for library computers to be triggered by an emag.
Things of note: I'm overriding some lists that get passed into buttons and one other thing using set, since the
list is alreadt generated by that step in the process? I think?
I've added dimness control to the dimmer component, since well, it was dimming already dark uis.
I also made and added a rather large background svg. I've got no experience with this sort of thing, and all the
compression methods I found for this ended up being busts. I know this isn't acceptable as an end product, but I
don't know how to get it there.

Somewhat on that note, this ui might not be worth the size for the amount of use it gets. I'm fine with nuking
it if that's the case, I bring this up because I have a very poor understanding of the logistics of something
like this, so I have a feeling I've fucked up somewhere

LemonInTheDark committed:
Forgot these, just a scss file for library computers, barely used but I think it's worthwhile

LemonInTheDark committed:
Missed this eariler. As a part of the uploading tab, I'm displaying the contents of books. I'm loading in that
context as raw html so paper -> book books look close to right. Means I need more html tags then our current
sanitize provides. I don't think any of these will cause issues, and there's also a good chance I'm missing
some. Will come up with a list later

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into book

LemonInTheDark committed:
Updates the rest of the maps to use the new management typepath

LemonInTheDark committed:
Fixes the default bible name being Default Bible Name, I am sorry

LemonInTheDark committed:
Turns out I had the scaling wrong for bible names, lead to weird stacking because the bible icon doesn't scale, so I lowered its sizing

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into book

LemonInTheDark committed:
Yeets unneeded exports (Thank you jlsnow)

Haha wouldn't it be funny if I didn't know how components worked

Co-authored-by: Jeremiah <42397676+jlsnow301@users.noreply.github.com>

LemonInTheDark committed:
merge master

LemonInTheDark committed:
Resets the maps to master

LemonInTheDark committed:
Fixes oversights from merge commit, changes maps

LemonInTheDark committed:
Removes needless Flex's from the scanner

LemonInTheDark committed:
Gives the library console the ability to parse markdown. Expands the list of acceptable html elements a bit

LemonInTheDark committed:
Adds audio cues for printing and inserting/removing from the scanner, makes the scanner nicer to use in general

LemonInTheDark committed:
Uses a compressed version of the background. It's still huge, but smaller at least

LemonInTheDark committed:
Adds the printing audio to the book binder

LemonInTheDark committed:
Merge branch 'book' of https://github.com/LemonInTheDark/tgstation into book

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into book

LemonInTheDark committed:
Cleans up tram

LemonInTheDark committed:
curse you tram

LemonInTheDark committed:
AHHHHHHHH

LemonInTheDark committed:
MY LIFE IS TRUE PAIN

LemonInTheDark committed:
Adds a path conversion statement to make people's lives easier

stylemistake commented on tgui/packages/tgui/interfaces/LibraryScanner.js:
```diff
@@ -0,0 +1,79 @@
+        </Section>
+      </Stack.Item>
+      <Stack.Item>
+        <Stack
+          width="100%">
```
```suggestion
        <Stack>
```

stylemistake commented on tgui/packages/tgui/interfaces/LibraryConsole.js:
```diff
@@ -0,0 +1,988 @@
+      title="Library Terminal"
+      width={880}
+      height={520}>
+      <PopoutMenu />
+      <Box fillPositionedParent left={show_dropdown ? "164px" : "50px"}>
+        <Window.Content>
+          <PageDisplay pb="10px" />
+        </Window.Content>
+      </Box>
```
Would it be easy for you to refactor this into stacks? Hard coded margins are bad.

LemonInTheDark replied:
I don't really want to end up in a position where the text overflows the space allotted, or gets too small in comparison. Is that reasonable? or am I thinking about this wrong

stylemistake replied:
Yes, it's possible to do in a nicer way. But if it's too hard for you, then it's ok to keep this as is, just make sure the gap is 6px between menu bar and the content.

LemonInTheDark replied:
![image](https://user-images.githubusercontent.com/58055496/157186963-6e24f8be-5d9e-46fc-b23c-2cc5fa2897d4.png)
What I want to do is something like this, which looks really nice, and dynamically accounts for string length, which is handy.
Trouble is what it does to modals. It makes them fill up to the whole window, rather then the tabs (for obvious reasons)
I can't nest the window content inside though, because if I do it just overflows without a hardcoded offset like that.

I'm partly doing this because I want to get better at js. I just don't know how I'd fix this without manually offsetting the window content, or not doing dimmers that way. Am happy to learn if it's possible, but I may be trying to do something that's inherently kinda crummy.

Regardless I'll ensure there's only a 6 px gap

LemonInTheDark replied:
Style helped me with this. I love you style

stylemistake commented on tgui/packages/tgui/interfaces/LibraryConsole.js:
```diff
@@ -0,0 +1,988 @@
+    screen_state,
+  } = data;
+
+  if (screen_state === 1) {
+    return (
+      <Inventory />
+    );
+  }
+  if (screen_state === 2) {
+    return (
+      <Checkout />
+    );
+  }
+  if (screen_state === 3) {
+    return (
+      <Archive />
+    );
+  }
+  if (screen_state === 4) {
+    return (
+      <Upload />
+    );
+  }
+  if (screen_state === 5) {
+    return (
+      <Print />
+    );
+  }
+  if (screen_state === 6) {
+    return (
+      <Forbidden />
+    );
+  }
```
For "cool syntax" points, add eslint-ignore if linter complains.

```suggestion
  return (
    screen_state === 1 ? <Inventory /> :
    screen_state === 2 ? <Checkout /> :
    screen_state === 3 ? <Archive /> :
    screen_state === 4 ? <Upload /> :
    screen_state === 5 ? <Print /> :
    screen_state === 6 ? <Forbidden /> :
    null
  );
```

stylemistake commented on tgui/packages/tgui/interfaces/LibraryConsole.js:
```diff
@@ -0,0 +1,988 @@
+  } = data;
+
+  if (!has_checkout) {
+    return;
```
Returning `null` is a good practice, since `undefined` normally means that you forgot to return something in the component. I know this because React catches these types of inaccuracies, and Inferno doesn't. Not very critical.

```suggestion
    return null;
```

stylemistake commented on tgui/packages/tgui/interfaces/LibraryConsole.js:
```diff
@@ -0,0 +1,988 @@
+    __html: run_marked_default(sanitizeText(cache_content)),
+  };
+  return (
+    <Box height="100%">
```
Suggesting to replace this with a fragment `<></>`, since modal doesn't need to be contained in a div, and fragments tags conveniently disappear in the final markup.

LemonInTheDark replied:
honestly didn't know that was possible. sure

stylemistake commented on tgui/packages/tgui/interfaces/LibraryConsole.js:
```diff
@@ -0,0 +1,988 @@
+            Current Scan Cache
+          </Box>
+        </Stack.Item>
+        <Stack.Item height="100%">
```
```suggestion
        <Stack.Item grow>
```

stylemistake commented on tgui/packages/tgui/interfaces/LibraryConsole.js:
```diff
@@ -0,0 +1,988 @@
+      <Stack.Item grow>
+        <Stack vertical height="100%">
+          <Stack.Item grow>
+            <Stack vertical height="100%">
```
Unnecessary stack, unwrap it.

LemonInTheDark replied:
I remember there being a reason for this, might have just been using stack wrong tho

LemonInTheDark replied:
Ah yeah uh.
I want the stack item to stick to the bottom of the window, if I just remove the second stack that breaks
![image](https://user-images.githubusercontent.com/58055496/157172415-f9f3bb4b-ec6e-4458-96e4-5732786498c5.png)

I think I could use mt="auto" for this, but that breaks when I resize the window, which I'm not a huge fan of

![image](https://user-images.githubusercontent.com/58055496/157172245-9791cf04-1f89-4622-b11a-cd518337552b.png)

Am I doing something wrong here? or is that overlap behavior just kinda how it should work

stylemistake replied:
You're doing something wrong, but ok.

LemonInTheDark replied:
Made better. Think I can go further with it, will work on it tomorrow

stylemistake commented:
@LemonInTheDark https://jakearchibald.github.io/svgomg/

LemonInTheDark committed:
Apply's style's suggestions

thx style

Co-authored-by: Aleksej Komarov <stylemistake@gmail.com>

LemonInTheDark committed:
Compresses the background svg

LemonInTheDark committed:
Further js cleanup

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into book

LemonInTheDark committed:
We no longer render markdown in the ui, since any source of markdown is converted to html anyway

cacogen commented:
finally the wgw machine receives some love

LemonInTheDark committed:
More ui changes

Makes the tab/main screen logic use Flex rather then manuel offsets
Makes modals better fit the size of their contents
Readjusts the width of some inputs
Properly uses the header prop for a table
Makes the buttons in the upload panel look nicer
Restructures the print tab a bit

LemonInTheDark committed:
Increase a modal's size

LemonInTheDark commented:
I'm going to add proper on/off overlays to these computers, because the lack of them has been tripping me up so hard

LemonInTheDark committed:
Fixes computers with no keyboard overlay showing their screen even when the power is out

LemonInTheDark commented:
It turns out they already had emissives, it's just the logic was borked so they weren't during off in darkness

Farquaar commented:
This is big. This could usher in a library renaissance.
I wonder if we'll have a feature to rate and/or tag books in the near future.

LemonInTheDark commented:
> This is big. This could usher in a library renaissance. I wonder if we'll have a feature to rate and/or tag books in the near future.

I've been planning to add a rating system, or at least a pass/fail system.
Once every like, 6 months the admin team has a panic attack about garbage in the library, wanted to add the base layer of "you need to get another player to support this" before a book goes from temporary to persistent. Across like, a week or something.
Hope was it would filter junk, and give people a reason to bully others into reading what they write

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into book

LemonInTheDark commented:
I'm touching every station in the codebase at the moment, not gonna actively keep the merge conflicts down until the pr's got a possibility of being merged

timothymtorres commented:
This is easy on the eyes.  Well done!

Kylerace commented on code/modules/library/lib_items.dm:
```diff
@@ -26,13 +28,46 @@
+
+	//Time to populate that list
+	var/list/books_in_area = GLOB.roundstart_books_by_area[area_type]
+	for(var/obj/item/book/book in contents)
```
why are you searching for a single typepath of books here when there are 3 parent typepaths for books

LemonInTheDark replied:
Because I only care about books with info datums, which bibles and cutouts don't have. Also because create_random_books doesn't generate anything but standard books.

Kylerace commented on code/modules/library/lib_items.dm:
```diff
@@ -2,6 +2,8 @@
 #define BOOKCASE_ANCHORED 1
 #define BOOKCASE_FINISHED 2
 
+GLOBAL_LIST_EMPTY(roundstart_books_by_area)
```
```suggestion
///contains information on book info datums by area type.
///associative list of the form: list(area typepath = list([/datum/book_info] instances inside that area))
GLOBAL_LIST_EMPTY(roundstart_books_by_area)
```
or something to that effect, we have enough undocumented global lists

LemonInTheDark replied:
Moved to the library subsystem, and documented a bit better

Kylerace commented on code/modules/library/lib_items.dm:
```diff
@@ -353,7 +443,7 @@
 	throw_range = 5
 	w_class = WEIGHT_CLASS_TINY
-	var/obj/machinery/computer/bookmanagement/computer //Associated computer - Modes 1 to 3 use this
+	var/obj/machinery/computer/libraryconsole/bookmanagement/computer //Associated computer - Modes 1 to 3 use this
```
```suggestion
	///Associated computer - Modes 1 to 3 use this
	var/obj/machinery/computer/libraryconsole/bookmanagement/computer 
```

LemonInTheDark replied:
God this code was horrible

Kylerace commented on code/modules/library/lib_machines.dm:
```diff
@@ -3,14 +3,39 @@
+#define DEFAULT_UPLOAD_CATAGORY "Fiction"
+GLOBAL_LIST_INIT(book_categories, list("Any", "Fiction", "Non-Fiction", "Adult", "Reference", "Religion"))
+GLOBAL_LIST_INIT(upload_categories, list("Fiction", "Non-Fiction", "Adult", "Reference", "Religion"))
+GLOBAL_VAR_INIT(default_book_category, "Any")
```
should be a define

LemonInTheDark replied:
Done

Kylerace reviewed (changes requested):
dm code looks good

LemonInTheDark commented on code/controllers/subsystem/library.dm:
```diff
@@ -0,0 +1,14 @@
+	load_shelves()
+
+/datum/controller/subsystem/library/proc/load_shelves()
+	for(var/obj/structure/bookcase/case_to_load as anything in shelves_to_load)
```
this is a good point, thank you

LemonInTheDark committed:
Moves some data and logic onto the library subsystem. Kyler's review

Fixes harddels held by the library scanner. Makes the scanner's buffer
actually do something

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into book

LemonInTheDark committed:
Makes book icon randomization a proc rather then just copypasta'd code

LemonInTheDark committed:
Removes the kilo library edit, the soul was removed

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into book

LemonInTheDark committed:
Damn you san (Fixes mapconflicts)

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into book

LemonInTheDark commented:
It turns out the library scanner's "buffer" has literally never functioned. I've fixed that

LemonInTheDark committed:
Pain

LemonInTheDark commented:
Ready*TM*

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into book
